documents/wg-application.md

```diff
@@ -0,0 +1,93 @@
+
+## What is the working group about?
+
+We want to work together on getting a solid foundation for writing CUDA code.
+Getting a safe and production-ready development experience on Stable Rust is our primary goal!
+
+Also, currently a major obstacle for developing GPGPU applications and algorithms in Rust is a lack of learning resources.
+We plan to solve the "documentation debt" with a broad range of tutorials, examples and references.
+
+## What is the working group not about?
+
+The WG is not focused on promoting or developing "standard" frameworks.
+Instead, we want to provide basic and reliable support of the feature and inspire the community to start using it.
+This should lead to experimenting with different approaches on how to use it and creating awesome tooling.
+
+## Is your WG long-running or temporary?
+
+In our current vision, the WG should live until we fulfill our goals.
+
+In the end, we hope the WG will evolve into another one to cover similar topics:
+to support other GPGPU platforms or to create higher-level frameworks to improve end-to-end experience based on community feedback.
+
+## What is your long-term vision?
+
+Having a reliable and safe CUDA development experience is our ultimate goal.
+This should include:
+
+* Getting `nvptx64-nvidia-cuda` to a [Tier 2](https://forge.rust-lang.org/platform-support.html) support state.
+* Test infrastructure on a real hardware and running the tests during Rust CI process.
+* Rich documentation, references, tutorials and examples.
+* A broad set of new compile errors, warnings and lints for SIMT execution model to avoid pitfalls and ensure code soundness.
+
+## How do you expect the relationship to the project be?
+
+The WG will be responsible for CUDA related issues and user requests.
+This includes [already reported issues](https://github.com/rust-lang/rust/issues?q=is%3Aopen+is%3Aissue+label%3AO-NVPTX) and those that will be opened due to the work made by the WG.
+An important aspect is that the WG will take care of the state of the `rust-ptx-linker` and will do its best to avoid blocking someone else's work (like it, unfortunately, happened in [rust#59752](https://github.com/rust-lang/rust/pull/59752)).
+
+### How do you want to establish accountability?
+
+For that purpose, we will publish an agenda and made decisions from our meetings.
+Once we achieve important milestones we plan to make public announces.
+
+## Which other WGs do you expect to have close contact with?
+
+We would like to cooperate with Core or Compiler Teams in discussions about safety in SIMT code.
+Additionally, it would be very important to discuss a strategy of reliable deployment of `rust-ptx-linker` (likely as a `rustup` component).
+
+On the other hand, proposed [Machine Learning WG](https://internals.rust-lang.org/t/enabling-the-formation-of-new-working-groups/10218/11) could leverage CUDA accelerated computing capabilities, so we can discuss their use cases.
+
```

Comment:
I'd say here that we expect to have most contact with Alex Crichton, which could inform the Core Team about our work if they deem it necessary. 

There are some open questions about the soundness of CUDA kernels and Rust's type system, which we will raise to the Language team when we deem appropriate (maybe link here to some of the open issues?). 

We have also secured access to a Power9+Volta GPGPU cluster from OSI for CI (I think @peterhj requested the access?) and we'd like to work with the infra team to enable rust-lang/rust to use it at some point, but initially we'd like to experiment with using it on a smaller component like `stdsimd` to gain experience with it.

Reply:
Sorry for the prolonged absence (due to a combo of work + travel). We got moved around a couple times but what we actually have access to now is a POWER8 + 2x Pascal (Tesla P100) machine via usual HPC batch scheduler setup (torque/PBS) via another POWER8 head node. They did install nvidia-docker on the compute node though, so the docker-based stdsimd tests should run on it fine. I had a custom CI-like setup that was working for single GPU nodes, though the torque/PBS stuff was the last blocker I was dealing with.

Reply:
@peterhj wonderful, thanks for an update!

> They did install nvidia-docker on the compute node though, so the docker-based stdsimd tests should run on it fine.

So, let's say, we *theoretically* can run CUDA-adjusted and resource-restricted [Rust Playground](https://github.com/integer32llc/rust-playground) there? I feel it will be extremely helpful for quick experiments.

Perhaps, could you please provide basic documentation about these things when you'll have time? It will be useful for people who never worked with HPC job schedulers (me at least :wink:).

Reply:
I suppose that you get ssh access to some front-end nodes, and the only way to access the nodes with the GPGPUs is to schedule jobs with Torque. 

I don't think they allow Playground like services for people to execute arbitrary code there, but even if they do, we would need to host the playground somewhere else, and each command (e.g. run), would need to ssh into the system, schedule a job with Torque, wait for the job to be scheduled, fetch the job logs, and report back. 

So you would probably be waiting a while for the results of a playground run there.